## ETL hoja victimas

Realizo un análisis exploratorio de los datos, con la finalidad de encontrar patrones que permitan generar información para la toma de decisiones en relación a disminuir las víctimas fatales.

### Importo las librerias necesarias para comenzar el proceso de ETL 

In [22]:
import numpy as np
import pandas as pd
import herramientas
import warnings
warnings.filterwarnings("ignore")

Comienzo mi Analisis Exploratorio con la hoja **VICTIMAS** de mi base de datos 'homicidios'

In [23]:
df_victimas=pd.read_excel('data/homicidios.xlsx',sheet_name='VICTIMAS')
df_victimas

ID_hecho      FECHA  AAAA  MM  DD                   ROL VICTIMA  \
0    2016-0001 2016-01-01  2016   1   1             CONDUCTOR    MOTO   
1    2016-0002 2016-01-02  2016   1   2             CONDUCTOR    AUTO   
2    2016-0003 2016-01-03  2016   1   3             CONDUCTOR    MOTO   
3    2016-0004 2016-01-10  2016   1  10             CONDUCTOR    MOTO   
4    2016-0005 2016-01-21  2016   1  21             CONDUCTOR    MOTO   
..         ...        ...   ...  ..  ..                   ...     ...   
712  2021-0092 2021-12-12  2021  12  12                PEATON  PEATON   
713  2021-0093 2021-12-13  2021  12  13  PASAJERO_ACOMPAÑANTE    MOTO   
714  2021-0094 2021-12-20  2021  12  20  PASAJERO_ACOMPAÑANTE    MOTO   
715  2021-0095 2021-12-30  2021  12  30             CONDUCTOR    MOTO   
716  2021-0096 2021-12-15  2021  12  15             CONDUCTOR    AUTO   

          SEXO EDAD  FECHA_FALLECIMIENTO  
0    MASCULINO   19  2016-01-01 00:00:00  
1    MASCULINO   70  2016-01-02 00:00:00  
2    MASCULINO   30  2016-01-03 00:00:00  
3    MASCULINO   18                   SD  
4    MASCULINO   29  2016-02-01 00:00:00  
..         ...  ...                  ...  
712   FEMENINO   50  2021-12-12 00:00:00  
713   FEMENINO   18  2021-12-18 00:00:00  
714   FEMENINO   43  2021-12-20 00:00:00  
715  MASCULINO   27  2022-01-02 00:00:00  
716  MASCULINO   60  2021-12-20 00:00:00  

[717 rows x 10 columns]

In [24]:
df_victimas.columns

Index(['ID_hecho', 'FECHA', 'AAAA', 'MM', 'DD', 'ROL', 'VICTIMA', 'SEXO',
       'EDAD', 'FECHA_FALLECIMIENTO'],
      dtype='object')

Podemos observar que la hoja 'VICTIMAS' consta de 10 columnas y 717 filas inicialmente. A continuacion se puede observar el 'Diccionario de datos' con el cual podremos observar el significado de cada columna de nuestra base de datos

Diccionario de datos:

- **ID_hecho**: identificador unico del siniestro
- **FECHA**: fecha en formato dd/mm/aaaa
- **AAAA**:	año
- **MM**: mes
- **DD**: día del mes
- **ROL**: Posición relativa al vehículo que presentaba la víctima en el momento del siniestro
- **VICTIMA**: Vehículo que ocupaba quien haya fallecido a se haya lastimado a raíz del hecho, o bien peatón/a. Clasificación agregada del tipo de vehículos.
    - **PEATON**:	Víctima distinta de cualquier ocupante de un vehículo, ya sea un conductor/a o un pasajero/a. Se incluyen los ocupantes o personas que empujan o arrastran un coche de bebé o una silla de ruedas o cualquier otro vehículo sin motor de pequeñas dimensiones. Se incluyen también las personas que caminan empujando una bicicleta o un ciclomotor.
    - **MOTO**:	Vehículo a motor no carrozado que incluye motocicleta, ciclomotor y cuatriciclo.
    - **AUTO**:	Vehículo a motor destinado al transporte de personas, diferente de los motovehículos, y que tenga hasta nueve plazas (incluyendo al asiento del conductor) (Sedan, SUV, coupe, etc)
    - **CARGAS**:	Vehículo a motor destiando al transporte de cargas, incluye camiones pesados (con o sin acoplado o semirremolque, etc., camión de recolección de residuos) y livianos (utilitarios, furgonetas, pick-ups, camioneta con caja de carga).
    - **BICICLETA**:	Vehículo con al menos dos ruedas, que generalmente es accionado por el esfuerzo muscular de las personas que lo ocupan, en particular mediante pedales o manivelas. Incluye bicicletas de pedaleo asistido y/o con motor
    - **PASAJEROS**:	Personas lesionadas que se encuentran dentro, descendiendo o ascendiendo de las unidades de autotrasporte público de pasajeros/as y ómnibus de larga distancia
    - **MOVIL**:	Vehículos de emergencia: móviles policiales, ambulancias, autobombas. 
    - **OTRO**:	otros vehiculos
    - **SD**:	Sin datos sobre el tipo de víctima.

- **SEXO**:	Sexo informado por fuente policial de la víctima
- **EDAD**:	Edad de la víctima al momento del siniestro
- **FECHA_FALLECIMIENTO**: Fecha de fallecimiento de la víctima


### A continuacion se procede a realizar el proceso de ETL

- **Estandarizacion del nombre de las columnas**: Dado que los nombres de las columnas algunos estan en minusculas, y el nombre de las mismas podria ser poco intuitivo a la hora de realizar analisis posteriormente, realizo la estandarizacion de los mismos

In [25]:
# Se coloca la primera en mayúscula
df_victimas.columns = [x.capitalize() for x in df_victimas.columns]
# Se reemplazan los guiones por espacios
df_victimas.columns = df_victimas.columns.str.replace('_', ' ')
# Se renombran algunas columnas
df_victimas = df_victimas.rename(columns={'Id hecho': 'Id',
                                                      'Aaaa':'Año',
                                                      'Mm':'Mes',
                                                      'Dd':'Dia',
                                                      'Victima':'Victima'})

Verificamos si se aplicaron los cambios anteriores

In [26]:
df_victimas.columns

Index(['Id', 'Fecha', 'Año', 'Mes', 'Dia', 'Rol', 'Victima', 'Sexo', 'Edad',
       'Fecha fallecimiento'],
      dtype='object')

Verifico valores nulos 

In [27]:
herramientas.verifica_tipo_y_nulos(df_victimas)

nombre_campo                                         tipo_datos  \
0                   Id                                    [<class 'str'>]   
1                Fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...   
2                  Año                                    [<class 'int'>]   
3                  Mes                                    [<class 'int'>]   
4                  Dia                                    [<class 'int'>]   
5                  Rol                                    [<class 'str'>]   
6              Victima                                    [<class 'str'>]   
7                 Sexo                                    [<class 'str'>]   
8                 Edad                     [<class 'int'>, <class 'str'>]   
9  Fecha fallecimiento       [<class 'datetime.datetime'>, <class 'str'>]   

   no_nulos_%  nulos_%  nulos  
0       100.0      0.0      0  
1       100.0      0.0      0  
2       100.0      0.0      0  
3       100.0      0.0      0  
4       100.0      0.0      0  
5       100.0      0.0      0  
6       100.0      0.0      0  
7       100.0      0.0      0  
8       100.0      0.0      0  
9       100.0      0.0      0

Verifico la existencia de valores duplicados

In [28]:
df_victimas.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
712    False
713    False
714    False
715    False
716    False
Length: 717, dtype: bool

Analizo por columna 'Id' si existen valores duplicados

In [29]:
herramientas.verifica_duplicados_por_columna(df_victimas,'Id')

Id      Fecha   Año  Mes  Dia                   Rol Victima  \
29   2016-0041 2016-03-29  2016    3   29             CONDUCTOR    MOTO   
30   2016-0041 2016-03-29  2016    3   29  PASAJERO_ACOMPAÑANTE    MOTO   
98   2016-0126 2016-09-18  2016    9   18             CONDUCTOR    AUTO   
99   2016-0126 2016-09-18  2016    9   18  PASAJERO_ACOMPAÑANTE    AUTO   
163  2017-0026 2017-02-26  2017    2   26  PASAJERO_ACOMPAÑANTE    AUTO   
164  2017-0026 2017-02-26  2017    2   26             CONDUCTOR    AUTO   
173  2017-0035 2017-03-23  2017    3   23             CONDUCTOR    AUTO   
174  2017-0035 2017-03-23  2017    3   23  PASAJERO_ACOMPAÑANTE    AUTO   
175  2017-0035 2017-03-23  2017    3   23  PASAJERO_ACOMPAÑANTE    AUTO   
177  2017-0036 2017-03-29  2017    3   29             CONDUCTOR    MOTO   
176  2017-0036 2017-03-29  2017    3   29  PASAJERO_ACOMPAÑANTE    MOTO   
186  2017-0050 2017-04-28  2017    4   28             CONDUCTOR    MOTO   
187  2017-0050 2017-04-28  2017    4   28  PASAJERO_ACOMPAÑANTE    MOTO   
237  2017-0108 2017-09-02  2017    9    2  PASAJERO_ACOMPAÑANTE    AUTO   
238  2017-0108 2017-09-02  2017    9    2  PASAJERO_ACOMPAÑANTE    AUTO   
242  2017-0112 2017-09-10  2017    9   10  PASAJERO_ACOMPAÑANTE    AUTO   
243  2017-0112 2017-09-10  2017    9   10  PASAJERO_ACOMPAÑANTE    AUTO   
253  2017-0126 2017-10-14  2017   10   14  PASAJERO_ACOMPAÑANTE    AUTO   
254  2017-0126 2017-10-14  2017   10   14             CONDUCTOR    AUTO   
278  2017-0153 2017-12-11  2017   12   11             CONDUCTOR    MOTO   
277  2017-0153 2017-12-11  2017   12   11  PASAJERO_ACOMPAÑANTE    MOTO   
300  2018-0015 2018-02-06  2018    2    6             CONDUCTOR    MOTO   
301  2018-0015 2018-02-06  2018    2    6  PASAJERO_ACOMPAÑANTE    MOTO   
312  2018-0026 2018-03-09  2018    3    9             CONDUCTOR   MOVIL   
313  2018-0026 2018-03-09  2018    3    9  PASAJERO_ACOMPAÑANTE   MOVIL   
325  2018-0038 2018-04-20  2018    4   20                PEATON  PEATON   
326  2018-0038 2018-04-20  2018    4   20                PEATON  PEATON   
373  2018-0085 2018-08-03  2018    8    3  PASAJERO_ACOMPAÑANTE    MOTO   
374  2018-0085 2018-08-03  2018    8    3             CONDUCTOR    MOTO   
388  2018-0098 2018-09-09  2018    9    9  PASAJERO_ACOMPAÑANTE    AUTO   
387  2018-0098 2018-09-09  2018    9    9             CONDUCTOR    AUTO   
418  2018-0128 2018-11-25  2018   11   25  PASAJERO_ACOMPAÑANTE    AUTO   
419  2018-0128 2018-11-25  2018   11   25  PASAJERO_ACOMPAÑANTE    AUTO   
444  2019-0010 2019-02-02  2019    2    2                PEATON  PEATON   
445  2019-0010 2019-02-02  2019    2    2                PEATON  PEATON   
576  2020-0038 2020-08-30  2020    8   30             CONDUCTOR    AUTO   
577  2020-0038 2020-08-30  2020    8   30             CONDUCTOR    AUTO   
602  2020-0063 2020-12-05  2020   12    5                PEATON  PEATON   
603  2020-0063 2020-12-05  2020   12    5  PASAJERO_ACOMPAÑANTE    MOTO   
616  2020-0077 2020-12-25  2020   12   25             CONDUCTOR    MOTO   
617  2020-0077 2020-12-25  2020   12   25  PASAJERO_ACOMPAÑANTE    MOTO   

          Sexo Edad  Fecha fallecimiento  
29   MASCULINO   54  2016-03-29 00:00:00  
30   MASCULINO   SD  2016-03-30 00:00:00  
98   MASCULINO   37                   SD  
99   MASCULINO   60                   SD  
163   FEMENINO   23  2017-02-26 00:00:00  
164  MASCULINO   19  2017-02-26 00:00:00  
173  MASCULINO   28  2017-03-23 00:00:00  
174  MASCULINO   32  2017-03-23 00:00:00  
175  MASCULINO   30  2017-03-23 00:00:00  
177  MASCULINO   20  2017-03-29 00:00:00  
176   FEMENINO   SD  2017-03-29 00:00:00  
186  MASCULINO   46  2017-04-28 00:00:00  
187  MASCULINO   16  2017-04-28 00:00:00  
237  MASCULINO   17  2017-09-02 00:00:00  
238  MASCULINO   10  2017-09-02 00:00:00  
242  MASCULINO    1                   SD  
243  MASCULINO   SD  2017-09-13 00:00:00  
253  MASCULINO   39                   SD  
254  MASCULINO   36  2017-10-14 00:00:00  
2

Se observan 'Id' repetidos pero varia el valor en la columna 'Rol' , por lo cual en un inicio se conservaran los valores repetidos en la columna 'Id'

Verifico el tipo de variable y los valores nulos

In [30]:
herramientas.verifica_tipo_y_nulos(df_victimas)

nombre_campo                                         tipo_datos  \
0                   Id                                    [<class 'str'>]   
1                Fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...   
2                  Año                                    [<class 'int'>]   
3                  Mes                                    [<class 'int'>]   
4                  Dia                                    [<class 'int'>]   
5                  Rol                                    [<class 'str'>]   
6              Victima                                    [<class 'str'>]   
7                 Sexo                                    [<class 'str'>]   
8                 Edad                     [<class 'int'>, <class 'str'>]   
9  Fecha fallecimiento       [<class 'datetime.datetime'>, <class 'str'>]   

   no_nulos_%  nulos_%  nulos  
0       100.0      0.0      0  
1       100.0      0.0      0  
2       100.0      0.0      0  
3       100.0      0.0      0  
4       100.0      0.0      0  
5       100.0      0.0      0  
6       100.0      0.0      0  
7       100.0      0.0      0  
8       100.0      0.0      0  
9       100.0      0.0      0

Se observa que la columna 'Edad' y 'Fecha fallecimiento' tienen varios tipos de datos, por lo cual se revisaran los mismos

- **Columna: 'Edad'**

In [31]:
# Cantidad de valores por tipo de dato en la columna 'edad'
datos_type_count = df_victimas['Edad'].apply(type).value_counts()
print('Cantidad de datos:')
print(datos_type_count)

Cantidad de datos:
Edad
<class 'int'>    664
<class 'str'>     53
Name: count, dtype: int64


Se observan 53 datos tipo str, se investigan los mismos 

In [32]:
df_victimas[df_victimas['Edad'].apply(lambda x: isinstance(x, str))]

Id      Fecha   Año  Mes  Dia                   Rol    Victima  \
30   2016-0041 2016-03-29  2016    3   29  PASAJERO_ACOMPAÑANTE       MOTO   
33   2016-0045 2016-04-11  2016    4   11             CONDUCTOR       MOTO   
35   2016-0048 2016-04-15  2016    4   15                PEATON     PEATON   
36   2016-0049 2016-04-17  2016    4   17                    SD         SD   
39   2016-0052 2016-04-20  2016    4   20                    SD       MOTO   
55   2016-0077 2016-06-13  2016    6   13                PEATON     PEATON   
63   2016-0085 2016-06-29  2016    6   29                    SD       MOTO   
72   2016-0096 2016-07-25  2016    7   25             CONDUCTOR       MOTO   
89   2016-0115 2016-09-02  2016    9    2                    SD         SD   
93   2016-0119 2016-09-04  2016    9    4  PASAJERO_ACOMPAÑANTE         SD   
108  2016-0136 2016-10-25  2016   10   25             CONDUCTOR       MOTO   
119  2016-0148 2016-11-16  2016   11   16             CONDUCTOR       MOTO   
121  2016-0151 2016-11-18  2016   11   18                PEATON     PEATON   
138  2016-0171 2016-12-25  2016   12   25             CONDUCTOR       MOTO   
141  2016-0174 2016-12-27  2016   12   27                    SD         SD   
142  2016-0175 2016-12-27  2016   12   27                PEATON     PEATON   
165  2017-0027 2017-02-26  2017    2   26                PEATON     PEATON   
176  2017-0036 2017-03-29  2017    3   29  PASAJERO_ACOMPAÑANTE       MOTO   
182  2017-0042 2017-04-10  2017    4   10             CONDUCTOR       MOTO   
225  2017-0093 2017-07-25  2017    7   25                PEATON     PEATON   
243  2017-0112 2017-09-10  2017    9   10  PASAJERO_ACOMPAÑANTE       AUTO   
275  2017-0151 2017-12-04  2017   12    4                PEATON     PEATON   
277  2017-0153 2017-12-11  2017   12   11  PASAJERO_ACOMPAÑANTE       MOTO   
299  2018-0014 2018-01-31  2018    1   31                PEATON     PEATON   
328  2018-0040 2018-04-27  2018    4   27                PEATON     PEATON   
359  2018-0071 2018-07-10  2018    7   10             CONDUCTOR       MOTO   
367  2018-0079 2018-07-23  2018    7   23                PEATON     PEATON   
371  2018-0083 2018-08-02  2018    8    2             CONDUCTOR       MOTO   
377  2018-0088 2018-08-17  2018    8   17             CONDUCTOR       MOTO   
391  2018-0101 2018-09-21  2018    9   21             CONDUCTOR       AUTO   
421  2018-0130 2018-12-03  2018   12    3             CONDUCTOR       MOTO   
444  2019-0010 2019-02-02  2019    2    2                PEATON     PEATON   
446  2019-0011 2019-02-10  2019    2   10                PEATON     PEATON   
454  2019-0019 2019-03-06  2019    3    6             CONDUCTOR       MOTO   
457  2019-0022 2019-03-15  2019    3   15  PASAJERO_ACOMPAÑANTE       AUTO   
461  2019-0026 2019-03-28  2019    3   28                PEATON     PEATON   
463  2019-0028 2019-04-06  2019    4    6                PEATON     PEATON   
469  2019-0034 2019-04-21  2019    4   21             CONDUCTOR       MOTO   
479  2019-0043 2019-05-21  2019    5   21              CICLISTA  BICICLETA   
498  2019-0063 2019-08-04  2019    8    4  PASAJERO_ACOMPAÑANTE       MOTO   
506  2019-0071 2019-08-27  2019    8   27                PEATON     PEATON   
529  2019-0094 2019-11-18  2019   11   18  PASAJERO_ACOMPAÑANTE       MOTO   
530  2019-0095 2019-11-20  2019   11   20             CONDUCTOR       MOTO   
548  2020-0010 2020-02-08  2020    2    8             CONDUCTOR       MOTO   
552  2020-0014 2020-02-27  2020    2   27             CONDUCTOR       MOTO   
555  2020-0017 2020-03-05  2020    3    5                PEATON     PEATON   
563  2020-0025 2020-05-12  2020    5   12                PEATON     PEATON   
572  2020-0034 2020-08-02  2020    8    2                PEATON     PEATON   
579  2020-0040 2020-09-12  2020    9   12             CONDUCTOR       AUTO   
631  2021-0012 2021-01-29  2021    1   29                PEATON     PEATON   
639  2021-0020 2021-02-21  2021   

Para completar los valores SD, se opta por realizar un promedio de las edades, teniendo en cuenta la columna 'sexo'. Como dicha columna tambien presenta valores SD, se procede a completar los mismos con el valor mas frecuente 

In [33]:
herramientas.valor_frecuente(df_victimas, 'Sexo')

El valor mas frecuente es: MASCULINO


Ahora se completa la edad , teninedo en cuenta el sexo

In [34]:
herramientas.edad_media_segun_sexo(df_victimas)

La edad promedio de Femenino es 51 y de Masculino es 40


- **Columnas: 'Rol' y 'Victima'**

Se observa ademas que las columnas 'Rol' y 'Victima' poseen faltantes, se revisan las cantidades de los mismos, es decir, valores SD

In [35]:
print(f" SD en 'Rol': {len(df_victimas[df_victimas['Rol']=='SD'])}")
print(f" SD en 'Victima': {len(df_victimas[df_victimas['Victima']=='SD'])}")

 SD en 'Rol': 11
 SD en 'Victima': 9


Se encontraron solo 11 en la columna 'Rol' y 9 en la columna 'Victima'. Por lo mismo reemplazamos dichos valores por el valor mas frecuente 

In [36]:
herramientas.valor_frecuente(df_victimas, 'Victima')

El valor mas frecuente es: MOTO


In [37]:
herramientas.valor_frecuente(df_victimas, 'Rol')

El valor mas frecuente es: CONDUCTOR


- **Columna: 'Fecha fallecimiento'**

Como el objetivo es analizar los datos para encontrar patrones que permitan tomar decisiones para disminuir los accidentes fatales, conocer la fecha de fallecimiento no se considera relevante para el mismo, por lo que se decide borrar la columna.

In [38]:
df_victimas = df_victimas.drop('Fecha fallecimiento', axis=1)
df_victimas

Id      Fecha   Año  Mes  Dia                   Rol Victima  \
0    2016-0001 2016-01-01  2016    1    1             CONDUCTOR    MOTO   
1    2016-0002 2016-01-02  2016    1    2             CONDUCTOR    AUTO   
2    2016-0003 2016-01-03  2016    1    3             CONDUCTOR    MOTO   
3    2016-0004 2016-01-10  2016    1   10             CONDUCTOR    MOTO   
4    2016-0005 2016-01-21  2016    1   21             CONDUCTOR    MOTO   
..         ...        ...   ...  ...  ...                   ...     ...   
712  2021-0092 2021-12-12  2021   12   12                PEATON  PEATON   
713  2021-0093 2021-12-13  2021   12   13  PASAJERO_ACOMPAÑANTE    MOTO   
714  2021-0094 2021-12-20  2021   12   20  PASAJERO_ACOMPAÑANTE    MOTO   
715  2021-0095 2021-12-30  2021   12   30             CONDUCTOR    MOTO   
716  2021-0096 2021-12-15  2021   12   15             CONDUCTOR    AUTO   

          Sexo  Edad  
0    MASCULINO    19  
1    MASCULINO    70  
2    MASCULINO    30  
3    MASCULINO    18  
4    MASCULINO    29  
..         ...   ...  
712   FEMENINO    50  
713   FEMENINO    18  
714   FEMENINO    43  
715  MASCULINO    27  
716  MASCULINO    60  

[717 rows x 9 columns]

Debido a que para los analisis posteriores, es necesario unir la informacion con el **df_hechos** , se procede a eliminar las columnas:'Fecha', 'Año', 'Mes', 'Dia' y 'Victima', debido a que es la misma que esta en el **df_hechos**.

In [39]:
df_hechos=pd.read_csv(r'data\hechos_limpio.csv')
df_hechos[df_hechos['Id']== '2016-0052']

Id  Cantidad de victimas       Fecha   Año  Mes  Dia      Hora  \
38  2016-0052                     1  2016-04-20  2016    4   20  20:00:00   

    Hora en entero             Lugar del hecho Tipo de calle  \
38              20  AUTOPISTA LUGONES PK 10000     AUTOPISTA   

                    Calle Cruce Dirección normalizada  Comuna XY (CABA)  \
38  LUGONES, LEOPOLDO AV.    No                    SD      13         0   

    Pos x  Pos y Participantes Victima Acusado  
38    0.0    0.0       MOTO-SD    MOTO      SD

In [40]:
df_victimas[df_victimas['Id']== '2016-0052']

Id      Fecha   Año  Mes  Dia        Rol Victima       Sexo  Edad
39  2016-0052 2016-04-20  2016    4   20  CONDUCTOR    MOTO  MASCULINO    39

In [41]:
df_victimas = df_victimas.drop(['Fecha', 'Año', 'Mes', 'Dia', 'Victima'], axis=1)
df_victimas.columns

Index(['Id', 'Rol', 'Sexo', 'Edad'], dtype='object')

### Carga de la Hoja 'VICTIMAS'

Se procede a guardar los cambios anteriormente realizados en un archivo llamado  **'victimas_limpio.csv'**

In [42]:
df_victimas_limpio = 'data/victimas_limpio.csv'
df_victimas.to_csv(df_victimas_limpio, index=False, encoding='utf-8')
print(f'Se guardó el archivo {df_victimas_limpio}')

Se guardó el archivo data/victimas_limpio.csv
